File: rust/13.罗马数字转整数.rs
/*
 * @lc app=leetcode.cn id=13 lang=rust
 *
 * [13] 罗马数字转整数
 */

// @lc code=start
impl Solution {
    pub fn roman_to_int(s: String) -> i32 {
        let s = s
            .replace("IV", "a")
            .replace("IX", "b")
            .replace("XL", "c")
            .replace("XC", "d")
            .replace("CD", "e")
            .replace("CM", "f");

        let mut result = 0;
        for c in s.chars() {
            result += match c {
                'I' => 1,
                'V' => 5,
                'X' => 10,
                'L' => 50,
                'C' => 100,
                'D' => 500,
                'M' => 1000,
                'a' => 4,
                'b' => 9,
                'c' => 40,
                'd' => 90,
                'e' => 400,
                'f' => 900,
                _ => 0,
            }
        }

        result
    }
}
// @lc code=end

File: 13.罗马数字转整数.go
/*
 * @lc app=leetcode.cn id=13 lang=golang
 *
 * [13] 罗马数字转整数
 */
package leetcode

// @lc code=start
func romanToInt(s string) int {
	romanMap := map[byte]int{'I': 1, 'V': 5, 'X': 10, 'L': 50, 'C': 100, 'D': 500, 'M': 1000}
	n := len(s)
	ans := romanMap[s[n-1]]
	for i := n - 2; i >= 0; i-- {
		if romanMap[s[i]] < romanMap[s[i+1]] {
			ans -= romanMap[s[i]]
		} else {
			ans += romanMap[s[i]]
		}
	}
	return ans
}

// @lc code=end
